src/components/ExpandCollapse/ExpandCollapse.tsx

```diff
@@ -0,0 +1,157 @@
+   */
+  initiallyExpanded?: boolean;
+  /**
+   * Defining this prop means that the component is controlled
+   *
+   * @default false
+   */
+  expanded?: boolean;
+  /**
+   * Change handler for the case when the component is controlled
+   */
+  onChange?: ReactEventHandler;
+  /**
+   * The duration for the appear transition
+   *
+   * @default 200
+   */
+  transitionDuration?: number;
+  /**
+   * A function accepting a boolean representing the current expansion state. Returns the
+   * collapsible/expandable content. Mutually exclusive with children.
+   */
+  content?: (x: boolean) => ReactElement;
+  /**
+   * A function accepting a boolean representing the current expansion state. Returns the
+   * collapsible/expandable content. Mutually exclusive with content.
+   */
+  children?: (x: boolean) => ReactElement;
+  dataTestId?: TestId;
+};
+
+function ExpandCollapse(props: Props) {
+  const {
+    textAndControl,
+    component = 'div',
+    transitionDuration = 200,
+    content,
+    children,
+    initiallyExpanded = false,
+    expanded: externallyControlledExpanded,
+    onChange,
+    dataTestId,
+  } = props;
+
+  if (
+    (externallyControlledExpanded === undefined && onChange) ||
+    (onChange === undefined && typeof externallyControlledExpanded === 'boolean')
+  ) {
+    throw new Error('If expanded is defined onChange must be defined too and vice versa');
+  }
+
+  const [internallyControlledExpanded, setInternallyControlledExpanded] = useState(
+    initiallyExpanded
+  );
+  const contentRef = useRef<HTMLDivElement>(null);
+
+  const Component = component;
+
+  const renderFunction = content ?? children;
+
+  if (renderFunction === undefined) {
+    throw new Error('Either content or children must be defined');
+  }
+
+  const expanded = externallyControlledExpanded ?? internallyControlledExpanded;
+
+  function handleStateChange(e: SyntheticEvent) {
+    if (typeof externallyControlledExpanded !== 'boolean') {
+      setInternallyControlledExpanded(state => !state);
+    } else {
+      if (!onChange) {
+        throw new Error('onChange function must be defined');
+      }
+      onChange(e);
+    }
+  }
+
+  function manageContentHeight() {
+    if (contentRef.current === null) {
+      throw new Error('Uninitialised element ref');
+    }
+
+    if (expanded) {
+      contentRef.current.style.height = ``;
+    } else {
+      contentRef.current.style.height = `0`;
+    }
+  }
+
+  function manageContentVisibility() {
+    if (contentRef.current === null) {
+      throw new Error('Uninitialised element ref');
+    }
+
+    let timeout: number;
+
+    if (expanded) {
+      contentRef.current.style.visibility = '';
+    } else {
+      timeout = window.setTimeout(() => {
+        if (contentRef.current === null) {
+          throw new Error('Uninitialised element ref');
+        }
+        contentRef.current.style.visibility = 'hidden';
+      }, transitionDuration);
+    }
+
+    return function() {
```

Comment:
also seeing a named function with this kind of return confused me for some time

Reply:
Non-issue